README.md

```diff
@@ -10,12 +10,93 @@ Basic interface for manifolds in Julia.
+
+### Mutating and allocating functions
+
+Every function, where this is applicable should provide a mutating and an allocating variant.
+For example for the exponential map `exp(M,p,x)` returns a _new_ point `q` where the result is computed in.
+On the other hand `exp!(M, q, p, X)` computes the result in place of `q`, where the design of the implementation
+should keep in mind that also `exp!(M,p,p,X)` should be possible without side effects.
+
+The interface provides a way to determine the allocation type and a result (see [`allocate_result`]()) to compute/allocate
+the resulting memory, such that the default implementation allocating functions, like `exp` is to allocate the resulting memory and call `exp!`.
+
+!!! note
+   it might be useful to provide two distinct implementations, for example when using AD schemes.
+   The default is meant for ease of use (concerning implementation), since then one has to implement the mutating variants.
+
+### The higher level interface and ease of use
+
+The higher level interface should aim for a convenience layer that resolves defaults and
+creates fallbacks for certain input parameters, that have these properties.
+It usually should not dispatch on a certain manifold nor on certain point or (co- or tangent) vector types.
+
+This layer should also not resolve/dispatch from the allocating to the mutating variant.
+
+This is maybe best illustrated by two examples
+
+1. the exponential map usually has a long form, where one can specify a fraction (of `X`) where the evaluation should be. This is generically impülemented a
+
+  ```julia
+    exp(M::AbstractManifold, p, X, t::Real) = exp(M, p, t * X)
+  ```
+
+  On this level neither the manifold _nor_ the points should be too strictly typed, points and vectors should – for best of cases – never be types.
+
+2. for the retraction, a default retraction (usually exp) is specified/defined via [`default_retraction_method`]().
+  This also means, that the last parameter of
+
+  ```julia
+  retract(M::AbstractManifold, p, X, ::AbstractRetractionMethod=default_retraction_method(M))
+  ```
+
+  is optional and for a concrete type the dispatch on a certain retraction is done next.
+  To avoid ambiguities, this concrete type should always be the first argument we dispatch on:
+  The `ExponentialRetractionMethod` calls `exp`, any other retraction calls a function of different name without this last parameter,
+  for example the `PolarRetractionMethod` by default calls `retract_polar(M,p,X)`, which is actualy a function from the lower level, see next section.
+
+### The lower level interface – performance
+
+This lower level aims for performance, that is, any function should have as less as possible optional and keyword arguments
+and be typed as concrete as possible/necessary. This means
+
+* the function name should be similar to its high level parent (for example `retract` and `retract_polar`from above)
+* The manifold should always be a concrete manifold
```

Comment:
```suggestion
* The manifold type in method signature should always be as narrow as possible.
```
The original wording may be confusing because there are concrete types in Julia and that's not what is meant here (f.e. `Euclidean` is not a concrete type but a perfectly fine argument type for a method).